android/libraries/rib-base/src/main/java/com/badoo/ribs/core/routing/transition/effect/SharedElementTransition.kt

```diff
@@ -0,0 +1,169 @@
+)
+
+@CheckResult
+fun <T> List<TransitionElement<out T>>.sharedElementTransition(
+    transitionParams: List<SharedElementTransition.Params>
+): Transition {
+    val exit = filter { it.direction == TransitionDirection.Exit }
+    val enter = filter { it.direction == TransitionDirection.Enter }
+    val transitions: MutableList<SharedElementTransitionInfo<T>> = mutableListOf()
+
+    transitionParams.forEach { transitionParam ->
+        var exitingView: View? = null
+        var enteringView: View? = null
+
+        val exitElementForId = exit.find {
+            exitingView = transitionParam.exitingElement.invoke(it.view)
+            exitingView != null
+        }
+
+        if (exitElementForId != null) {
+            val enteringElementForId = enter.find {
+                enteringView = transitionParam.enteringElement.invoke(it.view)
+                enteringView != null
+            }
+
+            if (enteringElementForId != null) {
+                transitions.add(
+                    SharedElementTransitionInfo(
+                        exitingElement = exitElementForId,
+                        exitingView = exitingView!!, // guaranteed by find clause
+                        enteringElement = enteringElementForId,
+                        enteringView = enteringView!!, // guaranteed by find clause
+                        params = transitionParam
+                    )
+                )
+            }
+        }
+    }
+
+    return Transition.multiple(
+        transitions.map {
+            it.transition()
+        }
+    )
+}
+
+@CheckResult
+internal fun <T> SharedElementTransitionInfo<T>.transition(): Transition {
+    val evaluator = SingleProgressEvaluator()
+    exitingElement.progressEvaluator.add(evaluator)
+    enteringElement.progressEvaluator.add(evaluator)
+
+    enteringElement.view.measure(0, 0)
+
+    val location = IntArray(2)
+    enteringView.getLocationInWindow(location)
+    val enteringAbsX = location[0]
+    val enteringAbsY = location[1]
+    exitingView.getLocationInWindow(location)
+    val exitingAbsX = location[0]
+    val exitingAbsY = location[1]
+    val exitingLayoutParams = (exitingView.layoutParams as ViewGroup.MarginLayoutParams)
+
+    val targetScaleX = 1.0f * enteringView.measuredWidth / exitingView.width
+    val targetScaleY = 1.0f * enteringView.measuredHeight / exitingView.height
+    val wDiff = (enteringView.measuredWidth - exitingView.width)
+    val hDiff = (enteringView.measuredHeight - exitingView.height)
+    val targetXDiff = enteringAbsX - exitingAbsX + wDiff / 2f
+    val targetYDiff = enteringAbsY - exitingAbsY + hDiff / 2f
+
+    fun Float.x(): Float = params.translateXInterpolator.getInterpolation(this)
+    fun Float.y(): Float = params.translateYInterpolator.getInterpolation(this)
+    fun Float.scaleX(): Float = params.scaleXInterpolator.getInterpolation(this)
+    fun Float.scaleY(): Float = params.scaleYInterpolator.getInterpolation(this)
```

Comment:
I can't imagine this implementation is useful for cases other than static context of the same size.
I can't use it for images, scale will ruin it.
I can't use it for text, scale will ruin it.

Looks like shared element transition is garbage too, but it least it is part of android framework. Can we somehow reuse it? Not `TransitionManager` directly, but `androidx.transition.ChangeBounds` directly for example. Looks like it is possible via `captureStartValues`, `captureEndValues` and `createAnimator`.
https://rubygarage.s3.amazonaws.com/uploads/article_image/file/3178/2_-_Tutorial_on_Shared_Element_Transitions_in_Android_Apps.gif

Reply:
Will look into it, thanks.
Any idea what to do with images and text?

Reply:
If you successfully reuse `androidx.transition`, then we can use `ChangeImageTransform` for images. Do not know, what is used for text, but probably there is some out of box solution.

Reply:
Usually they are crossfaded / scaled.